Source/WebCore/editing/SpellChecker.cpp

```diff
@@ -210,12 +228,44 @@ void SpellChecker::enqueueRequest(Ref<SpellCheckRequest>&& request)
     m_requestQueue.append(WTFMove(request));
 }
 
-void SpellChecker::didCheck(TextCheckingRequestIdentifier identifier, const Vector<TextCheckingResult>& results)
+static bool containsGrammarResult(TextCheckingResult result, const Vector<TextCheckingResult>& existingResults)
 {
-    ASSERT(m_processingRequest);
-    ASSERT(m_processingRequest->data().identifier() == identifier);
-    if (m_processingRequest->data().identifier() != identifier) {
-        m_requestQueue.clear();
+    bool foundIt = false;
+    for (TextCheckingResult existingResult : existingResults) {
+        TextCheckingType existingResultType = *existingResult.type.toSingleValue();
+        if (existingResultType != TextCheckingType::Grammar || result.range.location != existingResult.range.location || result.range.length != existingResult.range.length || result.details.size() != existingResult.details.size())
+            continue;
+        bool detailsMatch = true;
+        for (auto [detail, existingResultDetail] : zippedRange(result.details, existingResult.details)) {
+            if (!detailsMatch)
+                break;
+            detailsMatch = std::tie(detail.range.location, detail.range.length, detail.guesses) == std::tie(existingResultDetail.range.location, existingResultDetail.range.length, existingResultDetail.guesses);
+        }
+        if (detailsMatch)
+            foundIt = true;
+    }
+    return foundIt;
+}
+
+static bool containsAdditionalGrammarResults(const Vector<TextCheckingResult>& results, const Vector<TextCheckingResult>& existingResults)
+{
+    for (const auto& result : results) {
+        TextCheckingType resultType = *result.type.toSingleValue();
+        if (resultType == TextCheckingType::Grammar && !containsGrammarResult(result, existingResults))
```

Comment:
Same question about if we know for sure that `result.type` has exactly one bit set.

Maybe we can replace this in the same manner?

```suggestion
        if (result.type.containsOnly({ TextCheckingType::Grammar }) && !containsGrammarResult(result, existingResults))
```